File: signal-chat/crypto/ratchet.py
"""
Double Ratchet — per-message AES-256-GCM encryption with forward secrecy.
Uses dh.py for DH ratchet; cryptography lib only for AES-256-GCM and HKDF.
"""
import os
import json
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
from .dh import DHKeyPair


_KDF_RK_INFO = b"UIC-Signal-RatchetRoot-v1"
_KDF_CK_INFO = b"UIC-Signal-RatchetChain-v1"
_MAX_SKIP = 100


def _hkdf(ikm: bytes, salt: bytes, info: bytes, length: int = 64) -> bytes:
    return HKDF(
        algorithm=hashes.SHA256(),
        length=length,
        salt=salt,
        info=info,
    ).derive(ikm)


def _kdf_rk(root_key: bytes, dh_out: bytes) -> tuple[bytes, bytes]:
    out = _hkdf(dh_out, salt=root_key, info=_KDF_RK_INFO, length=64)
    return out[:32], out[32:]  # new_root_key, chain_key


def _kdf_ck(chain_key: bytes) -> tuple[bytes, bytes]:
    out = _hkdf(chain_key, salt=b"\x00" * 32, info=_KDF_CK_INFO, length=64)
    return out[:32], out[32:]  # new_chain_key, message_key


def _encrypt(mk: bytes, plaintext: bytes, aad: bytes = b"") -> bytes:
    nonce = os.urandom(12)
    ct = AESGCM(mk).encrypt(nonce, plaintext, aad)
    return nonce + ct


def _decrypt(mk: bytes, ciphertext: bytes, aad: bytes = b"") -> bytes:
    return AESGCM(mk).decrypt(ciphertext[:12], ciphertext[12:], aad)


class RatchetSession:
    """
    One Double Ratchet session between two parties.
    The 'initiator' flag distinguishes who sends first.

    For the receiver (initiator=False), pass their SPK DHKeyPair as
    `my_initial_keypair` so the first recv chain is derived correctly via
    DH(bob_spk_priv, alice_ratchet_pub).
    """

    def __init__(
        self,
        shared_key: bytes,
        their_ratchet_pub: int | None,
        initiator: bool,
        my_initial_keypair: DHKeyPair | None = None,
    ):
        self.root_key = shared_key
        self.sending_chain_key: bytes | None = None
        self.recv_chain_key: bytes | None = None
        # Sender uses a fresh ratchet key; receiver may reuse their SPK
        self.send_ratchet = my_initial_keypair if (not initiator and my_initial_keypair) else DHKeyPair()
        self.recv_ratchet_pub: int | None = their_ratchet_pub
        self.send_msg_num = 0
        self.recv_msg_num = 0
        self.prev_send_count = 0
        self.skipped: dict[tuple[int, int], bytes] = {}  # (pub, n) -> mk
        self.initiator = initiator

        if initiator and their_ratchet_pub is not None:
            # Sender: derive sending chain from DH(alice_ratchet, bob_spk)
            self.root_key, self.sending_chain_key = _kdf_rk(
                self.root_key,
                self.send_ratchet.dh(their_ratchet_pub),
            )
        elif not initiator and their_ratchet_pub is not None and my_initial_keypair is not None:
            # Receiver: derive recv chain from DH(bob_spk, alice_ratchet)
            self.root_key, self.recv_chain_key = _kdf_rk(
                self.root_key,
                my_initial_keypair.dh(their_ratchet_pub),
            )

    def encrypt(self, plaintext: str) -> dict:
        if self.sending_chain_key is None:
            raise RuntimeError("Sending chain not initialised")
        self.sending_chain_key, mk = _kdf_ck(self.sending_chain_key)
        header = {
            "dh": self.send_ratchet.public_key,
            "pn": self.prev_send_count,
            "n": self.send_msg_num,
        }
        self.send_msg_num += 1
        aad = json.dumps(header, sort_keys=True).encode()
        ct = _encrypt(mk, plaintext.encode(), aad)
        return {"header": header, "ciphertext": ct.hex()}

    def decrypt(self, message: dict) -> str:
        header = message["header"]
        ct = bytes.fromhex(message["ciphertext"])
        dh_pub = header["dh"]
        n = header["n"]
        pn = header["pn"]

        # Check skipped message keys first
        skip_key = (dh_pub, n)
        if skip_key in self.skipped:
            mk = self.skipped.pop(skip_key)
            aad = json.dumps(header, sort_keys=True).encode()
            return _decrypt(mk, ct, aad).decode()

        # DH ratchet step if public key changed
        if dh_pub != self.recv_ratchet_pub:
            self._skip_message_keys(pn)
            self._dh_ratchet(dh_pub)

        self._skip_message_keys(n)
        self.recv_chain_key, mk = _kdf_ck(self.recv_chain_key)
        self.recv_msg_num += 1
        aad = json.dumps(header, sort_keys=True).encode()
        return _decrypt(mk, ct, aad).decode()

    def _skip_message_keys(self, until: int):
        if self.recv_msg_num + _MAX_SKIP < until:
            raise RuntimeError("Too many skipped messages")
        while self.recv_chain_key and self.recv_msg_num < until:
            self.recv_chain_key, mk = _kdf_ck(self.recv_chain_key)
            self.skipped[(self.recv_ratchet_pub, self.recv_msg_num)] = mk
            self.recv_msg_num += 1

    def _dh_ratchet(self, their_pub: int):
        self.prev_send_count = self.send_msg_num
        self.send_msg_num = 0
        self.recv_msg_num = 0
        self.recv_ratchet_pub = their_pub
        self.root_key, self.recv_chain_key = _kdf_rk(
            self.root_key, self.send_ratchet.dh(their_pub)
        )
        self.send_ratchet = DHKeyPair()
        self.root_key, self.sending_chain_key = _kdf_rk(
            self.root_key, self.send_ratchet.dh(their_pub)
        )

    def get_ratchet_public(self) -> int:
        return self.send_ratchet.public_key

File: signal-chat/frontend/src/crypto/ratchet.js
// Double Ratchet — browser side.
// Mirrors crypto/ratchet.py exactly.
// DH ratchet uses dh.js (BigInt). AES-256-GCM + HKDF use WebCrypto API.

import { DHKeyPair, bigIntToHex, hexToBigInt, bytesToHex, hexToBytes } from "./dh.js";
import { hkdf } from "./x3dh.js";

const RK_INFO = new TextEncoder().encode("UIC-Signal-RatchetRoot-v1");
const CK_INFO = new TextEncoder().encode("UIC-Signal-RatchetChain-v1");
const MAX_SKIP = 100;

async function kdfRK(rootKey, dhOut) {
  const key = await crypto.subtle.importKey("raw", dhOut, { name: "HKDF" }, false, ["deriveBits"]);
  const bits = await crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt: rootKey, info: RK_INFO },
    key, 64 * 8
  );
  const out = new Uint8Array(bits);
  return [out.slice(0, 32), out.slice(32)];  // [newRootKey, chainKey]
}

async function kdfCK(chainKey) {
  const key = await crypto.subtle.importKey("raw", chainKey, { name: "HKDF" }, false, ["deriveBits"]);
  const bits = await crypto.subtle.deriveBits(
    { name: "HKDF", hash: "SHA-256", salt: new Uint8Array(32), info: CK_INFO },
    key, 64 * 8
  );
  const out = new Uint8Array(bits);
  return [out.slice(0, 32), out.slice(32)];  // [newChainKey, messageKey]
}

async function aesEncrypt(mk, plaintext, aad) {
  const key = await crypto.subtle.importKey("raw", mk, { name: "AES-GCM" }, false, ["encrypt"]);
  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const ct = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: nonce, additionalData: aad },
    key,
    new TextEncoder().encode(plaintext)
  );
  // nonce prepended, same layout as Python: nonce + ciphertext
  const out = new Uint8Array(12 + ct.byteLength);
  out.set(nonce);
  out.set(new Uint8Array(ct), 12);
  return out;
}

async function aesDecrypt(mk, ciphertext, aad) {
  const key = await crypto.subtle.importKey("raw", mk, { name: "AES-GCM" }, false, ["decrypt"]);
  const pt = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: ciphertext.slice(0, 12), additionalData: aad },
    key,
    ciphertext.slice(12)
  );
  return new TextDecoder().decode(pt);
}

export class RatchetSession {
  constructor() {
    this.rootKey         = null;
    this.sendingChainKey = null;
    this.recvChainKey    = null;
    this.sendRatchet     = new DHKeyPair();
    this.recvRatchetPub  = null;  // BigInt
    this.sendMsgNum      = 0;
    this.recvMsgNum      = 0;
    this.prevSendCount   = 0;
    this.skipped         = {};    // "pubHex:n" -> Uint8Array mk
  }

  // Alice calls this after x3dhSender
  // sharedKey: Uint8Array (32 bytes from X3DH)
  // theirRatchetPub: BigInt (Bob's SPK public key used as initial ratchet target)
  async initSender(sharedKey, theirRatchetPub) {
    this.recvRatchetPub = theirRatchetPub;
    const dhOut = this.sendRatchet.dh(theirRatchetPub);
    [this.rootKey, this.sendingChainKey] = await kdfRK(sharedKey, dhOut);
  }

  // Bob calls this after x3dhReceiver
  // sharedKey: Uint8Array (32 bytes from X3DH)
  // theirRatchetPub: BigInt (Alice's ratchet pub from message header)
  // myInitialKeypair: DHKeyPair (Bob's SPK — used to derive first recv chain)
  async initReceiver(sharedKey, theirRatchetPub, myInitialKeypair) {
    this.sendRatchet    = myInitialKeypair;
    this.recvRatchetPub = theirRatchetPub;
    const dhOut = myInitialKeypair.dh(theirRatchetPub);
    [this.rootKey, this.recvChainKey] = await kdfRK(sharedKey, dhOut);
  }

  async encrypt(plaintext) {
    if (!this.sendingChainKey) {
      // Bob's first send after receiving — derive sending chain with a fresh ratchet key.
      // Does not touch recvChainKey so Bob can still decrypt future messages from Alice.
      if (!this.recvRatchetPub) throw new Error("Sending chain not initialised");
      this.sendRatchet = new DHKeyPair();
      const dhOut = this.sendRatchet.dh(this.recvRatchetPub);
      [this.rootKey, this.sendingChainKey] = await kdfRK(this.rootKey, dhOut);
      this.sendMsgNum = 0;
      this.prevSendCount = 0;
    }
    const [newCK, mk] = await kdfCK(this.sendingChainKey);
    this.sendingChainKey = newCK;

    const header = {
      dh: bigIntToHex(this.sendRatchet.publicKey),
      pn: this.prevSendCount,
      n:  this.sendMsgNum,
    };
    this.sendMsgNum++;

    const aad = new TextEncoder().encode(JSON.stringify(header));
    const ct  = await aesEncrypt(mk, plaintext, aad);
    return { header, ciphertext: bytesToHex(ct) };
  }

  async decrypt(message) {
    const { header } = message;
    const ct = hexToBytes(message.ciphertext);
    const dhPub = hexToBigInt(header.dh);
    const n  = header.n;
    const pn = header.pn;

    // Check skipped message keys
    const skipKey = `${header.dh}:${n}`;
    if (this.skipped[skipKey]) {
      const mk = this.skipped[skipKey];
      delete this.skipped[skipKey];
      const aad = new TextEncoder().encode(JSON.stringify(header));
      return await aesDecrypt(mk, ct, aad);
    }

    const currentDHHex = this.recvRatchetPub ? bigIntToHex(this.recvRatchetPub) : null;
    if (header.dh !== currentDHHex) {
      await this._skipMessageKeys(pn);
      await this._dhRatchet(dhPub);
    }

    await this._skipMessageKeys(n);
    const [newCK, mk] = await kdfCK(this.recvChainKey);
    this.recvChainKey = newCK;
    this.recvMsgNum++;

    const aad = new TextEncoder().encode(JSON.stringify(header));
    return await aesDecrypt(mk, ct, aad);
  }

  async _skipMessageKeys(until) {
    if (this.recvMsgNum + MAX_SKIP < until) throw new Error("Too many skipped messages");
    while (this.recvChainKey && this.recvMsgNum < until) {
      const [newCK, mk] = await kdfCK(this.recvChainKey);
      this.recvChainKey = newCK;
      this.skipped[`${bigIntToHex(this.recvRatchetPub)}:${this.recvMsgNum}`] = mk;
      this.recvMsgNum++;
    }
  }

  async _dhRatchet(theirPub) {
    this.prevSendCount = this.sendMsgNum;
    this.sendMsgNum    = 0;
    this.recvMsgNum    = 0;
    this.recvRatchetPub = theirPub;
    const dh1 = this.sendRatchet.dh(theirPub);
    [this.rootKey, this.recvChainKey] = await kdfRK(this.rootKey, dh1);
    this.sendRatchet = new DHKeyPair();
    const dh2 = this.sendRatchet.dh(theirPub);
    [this.rootKey, this.sendingChainKey] = await kdfRK(this.rootKey, dh2);
  }

  // ── Persistence ───────────────────────────────────────────────

  serialize() {
    const skippedSer = {};
    for (const [k, v] of Object.entries(this.skipped))
      skippedSer[k] = bytesToHex(v);
    return JSON.stringify({
      rootKey:         this.rootKey         ? bytesToHex(this.rootKey)         : null,
      sendingChainKey: this.sendingChainKey ? bytesToHex(this.sendingChainKey) : null,
      recvChainKey:    this.recvChainKey    ? bytesToHex(this.recvChainKey)    : null,
      sendRatchet:     this.sendRatchet.toJSON(),
      recvRatchetPub:  this.recvRatchetPub  ? bigIntToHex(this.recvRatchetPub) : null,
      sendMsgNum:      this.sendMsgNum,
      recvMsgNum:      this.recvMsgNum,
      prevSendCount:   this.prevSendCount,
      skipped:         skippedSer,
    });
  }

  static deserialize(json) {
    const d = JSON.parse(json);
    const sess = new RatchetSession();
    sess.rootKey         = d.rootKey         ? hexToBytes(d.rootKey)         : null;
    sess.sendingChainKey = d.sendingChainKey ? hexToBytes(d.sendingChainKey) : null;
    sess.recvChainKey    = d.recvChainKey    ? hexToBytes(d.recvChainKey)    : null;
    sess.sendRatchet     = DHKeyPair.fromJSON(d.sendRatchet);
    sess.recvRatchetPub  = d.recvRatchetPub  ? hexToBigInt(d.recvRatchetPub) : null;
    sess.sendMsgNum      = d.sendMsgNum;
    sess.recvMsgNum      = d.recvMsgNum;
    sess.prevSendCount   = d.prevSendCount;
    const skipped = {};
    for (const [k, v] of Object.entries(d.skipped || {}))
      skipped[k] = hexToBytes(v);
    sess.skipped = skipped;
    return sess;
  }
}
